File: array/problem154/problem154.py
from typing import List


class Solution:
    def findMin(self, nums: List[int]) -> int:
        # With duplicates, worst-case can degrade to O(n) when boundaries are equal
        lo, hi = 0, len(nums) - 1
        while lo < hi:
            mid = (lo + hi) // 2
            if nums[mid] < nums[hi]:
                hi = mid
            elif nums[mid] > nums[hi]:
                lo = mid + 1
            else:
                # nums[mid] == nums[hi], cannot decide side; shrink safely
                hi -= 1
        return nums[lo]

File: array/problem154/problem154.cpp
#include <vector>

using namespace std;

class Solution {
public:
    int findMin(vector<int>& nums) {
        // With duplicates, worst-case can degrade to O(n)
        int lo = 0, hi = static_cast<int>(nums.size()) - 1;
        while (lo < hi) {
            int mid = lo + (hi - lo) / 2;
            if (nums[mid] < nums[hi]) {
                hi = mid;
            } else if (nums[mid] > nums[hi]) {
                lo = mid + 1;
            } else {
                // nums[mid] == nums[hi] -> shrink right boundary
                --hi;
            }
        }
        return nums[lo];
    }
};
